### PR TITLE
Display a warning when there are multiple documents or folders of the same name

### PR DESCRIPTION
**Pull request by jonpo**
@isaacwhite 
### Description of Change
Add handling to detect duplicate slugs in `buildTreeFromData`. Flag duplicate resources and use that flag to display an error message when rendering a response. This doesn't address the problem where if there are duplicates Library doesn't consistently show one of them––wasn't sure the best way to go about that. Let me know if there's a better way of doing this that I didn't see!

Because of the way file fixtures are being generated, adding to them (in order to test the duplicate scenario) changed some fixture identifiers, which is why I had to update so many tests here. It might be worth changing the way those fixtures get generated so that adding to them in the future doesn't change the identifiers?

### Related Issue
https://github.com/nytimes/library/issues/4

### Screenshots
Screen Shot 2019-07-16 at 5 06 28 PM
Screen Shot 2019-07-17 at 1 08 44 PM

### Motivation and Context
Let Library users know that the contents of a folder are not accurately represented in Library if there are multiple documents or folders of the same name.

### Checklist

- [x] Ran `npm run lint` and updated code style accordingly
- [x] `npm run test` passes
- [x] PR has a description and all contributors/stakeholder are noted/cc'ed
- [x] tests are updated and/or added to cover new code
- [ ] relevant documentation is changed and/or added

### CODE REVIEW

File: layouts/partials/warning.ejs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<div class="warning">
+  <%= message %>
```
Comment (isaacwhite) [inline layouts/partials/warning.ejs]: Since this template is named `warning`, would it make sense to include the `Warning:` as part of the template?

File: server/list.js [hunk 2]
```diff
@@ -235,8 +235,14 @@ function buildTreeFromData(rootParent, previousData, breadcrumb) {
     const {slug} = docsInfo[id]
     const nextCrumb = breadcrumb ? breadcrumb.concat({ id: rootParent, slug: parentInfo.slug }) : []
 
-    // recurse building up breadcrumb
-    memo.children[slug] = buildTreeFromData(id, previousData, nextCrumb)
+    if (!memo.children[slug]) {
+      // recurse building up breadcrumb
+      memo.children[slug] = buildTreeFromData(id, previousData, nextCrumb)
+    } else {
+      log.debug(`Folder ${parentInfo.name} contains duplicate resources with slug ${slug}`)
```
Comment (isaacwhite) [inline server/list.js]: How would you feel about this as a `warn` level log, since we shouldn't really encourage this case to persist?

File: config/strings.yaml [hunk 3]
```diff
@@ -80,3 +80,8 @@ error:
     heading: Uh oh, that's a 500.
     message: Perhaps try to search for something else instead?
 
+# warning messages
+warning:
+  duplicateDoc: "Warning: Multiple documents in this folder share the same name as this document. Only one of the documents will be accesible through Library."
```
Comment (isaacwhite) [inline config/strings.yaml]: It looks like we don't know for sure whether the other item in the drive that has the exact same slug as this item is the same type; you could mix and match.

A little more complex than a static warning message, but what if we made the warning include the names of the other items that would have shown up at this location to make it easier to resolve the issue? We could save the names of the other documents (or folders!) where we currently set  `isDuplicate`, then pass them into a yaml function like in the `moveFile` definition. In that case, we might want to discard the boolean and instead create an array called `duplicates` as a property.

The general logic of the diff wouldn't change much, but we could drop the ternary in `nav.ejs` and be more specific without needing as many configuration options in strings.yaml.

File: layouts/partials/warning.ejs (hunk 1 above)
Comment (jonpo) [reply layouts/partials/warning.ejs]: 👍

File: server/list.js (hunk 2 above)
Comment (jonpo) [reply server/list.js]: 👍

File: config/strings.yaml (hunk 3 above)
Comment (jonpo) [reply config/strings.yaml]: Good idea! I'll make the changes.

**Comment by jonpo** [comment]
@isaacwhite I addressed your comments. Now the error message looks like this:
```Warning: Multiple resources in this folder share the same name: duplicate document, duplicate document. Only one will be accesible through Library.```
Do you think it would be more helpful if we displayed the names as links to the actual drive resources?

File: server/routes/categories.js [hunk 4]
```diff
@@ -44,7 +44,8 @@ async function handleCategory(req, res) {
     createdAt: moment(meta.createdTime).fromNow(),
     editLink: meta.mimeType === 'text/html' ? meta.folder.webViewLink : meta.webViewLink,
     id,
-    template: stringTemplate
+    template: stringTemplate,
+    duplicateNames: duplicates ? duplicates.join(', ') : ''
```
Comment (isaacwhite) [inline server/routes/categories.js]: What do you think about passing the `duplicates` directly into the render call so that future changes wouldn't touch this file? Then we could perform the `.join()` inside the strings.yaml itself.

File: server/list.js [hunk 5]
```diff
@@ -235,8 +235,17 @@ function buildTreeFromData(rootParent, previousData, breadcrumb) {
...
-    // recurse building up breadcrumb
-    memo.children[slug] = buildTreeFromData(id, previousData, nextCrumb)
+    if (!memo.children[slug]) {
+      // recurse building up breadcrumb
+      memo.children[slug] = buildTreeFromData(id, previousData, nextCrumb)
+    } else {
+      log.warn(`Folder ${parentInfo.name} contains duplicate resources with slug ${slug}`)
+      const { name } = docsInfo[id]
+      // record duplicate resources so that a warning can be displayed
+      memo.children[slug].duplicates = memo.children[slug].duplicates
```
Comment (isaacwhite) [inline server/list.js]: Could go either way, but would this be more explicit?
```
const previousDupes = memo.children[slug].duplicates || []
memo.children[slug].duplicates = previousDupes.concat(name)
```

File: config/strings.yaml [hunk 6]
```diff
@@ -80,3 +80,7 @@ error:
     heading: Uh oh, that's a 500.
     message: Perhaps try to search for something else instead?
 
+# warning messages
+warning:
+  duplicate: !!js/function (names) => `Multiple resources in this folder share the same name${':' + ' ' + names}. Only one will be accesible through Library.`
```
[the remark this answers is not in the document]
Comment (jonpo) [reply config/strings.yaml]: That's what I originally had, but I was getting this error:
```
[0]   throw generateError(state, message);
[0]   ^
[0] YAMLException: cannot resolve a node with !<tag:yaml.org,2002:js/function> explicit tag at line 85, column 93:
[0]      ...  this folder share the same name: ${names}. Only one will be acc ...
```
Tried googling around a bit but I didn't find anything useful. What I have here is the only variation I tried that didn't throw the error 🙁

File: server/routes/categories.js (hunk 4 above)
Comment (jonpo) [reply server/routes/categories.js]: 👍

File: server/list.js (hunk 5 above)
Comment (jonpo) [reply server/list.js]: Yeah I like that 👍

File: config/strings.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (isaacwhite) [reply config/strings.yaml]: Ah, I see. It's getting upset about a `:` character as part of a key not enclosed in quotes. Two other options I found:
```
... this folder share the same name&#58; ${names}...
```
where the character sequence doesn't trip up the parser

or
```
... this folder share the same name${":"} ${names}...
```
which might be only slightly better because it's more obvious the problem with the `:` character.

But I could go either way, it's unfortunate that the parser doesn't interpret the backticks as quotes the same way.

What do you want to do?

[the remark this answers is not in the document]
Comment (jonpo) [reply config/strings.yaml]: html entity is the best option I think
